### PR TITLE
Adding swap file to loader nodes during setup process #1118

### PR DESCRIPTION
**Pull request by aleksbykov**
## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/cluster.py [hunk 1]
```diff
@@ -3372,6 +3374,21 @@ def kill_gemini_thread(self):
                 if kill_result.exit_status != 0:
                     self.log.error('Terminate gemini on node %s:\n%s', loader, kill_result)
 
+    def create_swap_file(self):
+        swap_size = self.params.get("loader_swap_size", default=None)
+        if not swap_size:
+            self.log.info("Swap file is not configured")
+            return
+        commands = dedent("""sudo /bin/dd if=/dev/zero of=/var/swap.1 bs=1M count={}
```
Comment (fruch) [inline sdcm/cluster.py]: Those commands were tested on Ubuntu/Debian ?

File: sdcm/cluster.py [hunk 2]
```diff
@@ -3372,6 +3374,21 @@ def kill_gemini_thread(self):
...
+                          sudo /sbin/mkswap /var/swap.1
```
Comment (amoskong) [inline sdcm/cluster.py]: Let's use more meaningful name -- like sct_configured_swapfile?

File: sdcm/cluster.py [hunk 3]
```diff
@@ -3372,6 +3374,21 @@ def kill_gemini_thread(self):
...
+                          sudo chmod 600 /var/swap.1
+                          sudo /sbin/swapon /var/swap.1""".format(swap_size))
```
Comment (amoskong) [inline sdcm/cluster.py]: how about adding a grep statement to make sure if swap file is using?
grep '/var/swap.1' /proc/swaps

File: sdcm/cluster.py [hunk 4]
```diff
@@ -3045,6 +3045,8 @@ def install_gemini(self, node):
     def node_setup(self, node, verbose=False, db_node_address=None, **kwargs):  # pylint: disable=unused-argument
         self.log.info('Setup in BaseLoaderSet')
         node.wait_ssh_up(verbose=verbose)
+        # add swap file
+        self.create_swap_file()
```
Comment (amoskong) [inline sdcm/cluster.py]: Do we need to skip this for REUSE cluster? or checking if swap is configured before creating swap file?

File: defaults/test_default.yaml [hunk 5]
```diff
@@ -48,3 +48,5 @@ send_email: false
 email_recipients: ['qa@scylladb.com']
 
 collect_logs: false
+
+loader_swap_size: 1024 # Size of swap file for loader node in bytes 1024 * 1MB
```
Comment (amoskong) [inline defaults/test_default.yaml]: 1G should be not enough for the default loader?
Can we increase it to be same as the memory size of default loader instance type?

File: sdcm/cluster.py [hunk 6]
```diff
@@ -3372,6 +3374,21 @@ def kill_gemini_thread(self):
                 if kill_result.exit_status != 0:
                     self.log.error('Terminate gemini on node %s:\n%s', loader, kill_result)
 
+    def create_swap_file(self):
```
Comment (amoskong) [inline sdcm/cluster.py]: Can you move this common function to base node, we can also use that for db node & monitor?

File: sdcm/cluster.py (hunk 1 above)
Comment (amoskong) [reply sdcm/cluster.py]: It's worth to test and make it work.

Currently we only use centos7 for loader & monitor?

**Comment by fruch** [comment]
@aleksbykov, can you put the trello link in the top of the PR message ? so we can get the context of this PR ?

File: sdcm/cluster.py (hunk 4 above)
Comment (fruch) [reply sdcm/cluster.py]: We should skip for REUSE, reuse assumes everything configured already.

File: sdcm/cluster.py (hunk 1 above)
Comment (fruch) [reply sdcm/cluster.py]: if it's not working on debian/ubuntu, we should put an if here and raise exception

File: sdcm/cluster.py (hunk 2 above)
Comment (fruch) [reply sdcm/cluster.py]: yeah, would be nice to have a clear name

File: sdcm/cluster.py (hunk 6 above)
Comment (fruch) [reply sdcm/cluster.py]: not critical, once we decided we need it for the db nodes, we can move it to a better place

File: defaults/test_default.yaml (hunk 5 above)
Comment (fruch) [reply defaults/test_default.yaml]: A good idea, we should try to put enough memory, cause we are sure not missing only 1Gb

Comment (bentsi) [reply defaults/test_default.yaml]: usually the recommended swap is the same as a memry size. Check how muchwe can increase this (but not increasing the root disk size)

File: sdcm/cluster.py (hunk 6 above)
Comment (bentsi) [reply sdcm/cluster.py]: move to BaseNode

File: sdcm/cluster.py (hunk 3 above)
Comment (bentsi) [reply sdcm/cluster.py]: agree

File: sdcm/sct_config.py [hunk 7]
```diff
@@ -768,7 +768,10 @@ class SCTConfiguration(dict):
              """,
              choices=("keep", "keep-on-failure", "destroy")),
         dict(name="workaround_kernel_bug_for_iotune", env="SCT_WORKAROUND_KERNEL_BUG_FOR_IOTUNE", type=bool,
-             help="Workaround a known kernel bug which causes iotune to fail in scylla_io_setup, only effect GCE backend")
+             help="Workaround a known kernel bug which causes iotune to fail in scylla_io_setup, only effect GCE backend"),
+
+        dict(name="loader_swap_size", env="SCT_LOADER_SWAP_SIZE", type=int,
+             help="Swap file size in bytes calculated by x * 1MB")
```
Comment (bentsi) [inline sdcm/sct_config.py]: put a comma after your line so when next param will be added one will not need change unnecessary code

**Comment by fruch** [comment]
@aleksbykov this one is waiting for you

File: sdcm/cluster.py (hunk 2 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: changed!

File: sdcm/cluster.py (hunk 1 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: checked on 18.04. steps are working

File: sdcm/cluster.py (hunk 3 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: Thanks. added

File: defaults/test_default.yaml (hunk 5 above)
Comment (aleksbykov) [reply defaults/test_default.yaml]: @bentsi @amoskong @fruch the default loader instance memory is 8GB, but disk size looks 10GB. if Create swap equal to memory size, the 2GB of disk will not be enough, i think

File: sdcm/cluster.py (hunk 6 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: moved

File: sdcm/sct_config.py (hunk 7 above)
Comment (aleksbykov) [reply sdcm/sct_config.py]: ok.

File: defaults/test_default.yaml (hunk 5 above)
Comment (aleksbykov) [reply defaults/test_default.yaml]: # df -h
Filesystem      Size  Used Avail Use% Mounted on
devtmpfs        1,8G     0  1,8G   0% /dev
tmpfs           1,9G     0  1,9G   0% /dev/shm
tmpfs           1,9G   17M  1,9G   1% /run
tmpfs           1,9G     0  1,9G   0% /sys/fs/cgroup
/dev/nvme0n1p1   10G  5,9G  4,1G  60% /
tmpfs           374M     0  374M   0% /run/user/1000

# cat /proc/meminfo
MemTotal:        3822536 kB

Comment (fruch) [reply defaults/test_default.yaml]: maybe we could enlarge the default diskspace ? we are doing that for the other type of nodes already
Disk isn't being use that much in the loaders, if i'm not mistaken.

```
aws_root_disk_size_monitor: 35  # GB, remove this field if default disk size should be used
aws_root_disk_size_db: 15
```

File: sdcm/cluster.py (hunk 1 above)
Comment (fruch) [reply sdcm/cluster.py]: good

File: defaults/test_default.yaml (hunk 5 above)
Comment (fruch) [reply defaults/test_default.yaml]: @bentsi what do you say we push this like that, and open a new PR for enlarging the disksize of the loaders, to accommodate bigger swap ?

Comment (bentsi) [reply defaults/test_default.yaml]: yes, let's keep like this 
@aleksbykov please open a task in Trello to increase the loader disk size

**Comment by bentsi** [comment]
Not backporing this, since this doesn't fix something critical.

File: defaults/test_default.yaml (hunk 5 above)
Comment (aleksbykov) [reply defaults/test_default.yaml]: Done: https://trello.com/c/A9gkrM0z

**Pull request merged**